### ISSUE TITLE
Internal Compiler Error (codegen): "there was an error in verifying the lgenfe output!"

### ISSUE THREAD

**Issue by pauljurczak** [kokkos/kokkos#3496]
I'm building Kokkos on Ubuntu 18.04.5 with gcc 8.4.0 and Cuda compilation tools, release 11.0, V11.0.194. Here is my configuration:
```
Setting policy CMP0074 to use <Package>_ROOT variables
The project name is: Kokkos
Compiler Version: 11.0.194
Using -std=c++17 for C++17 standard as feature
Execution Spaces:
    Device Parallel: CUDA
    Host Parallel: OPENMP
      Host Serial: SERIAL

Architectures:
 KEPLER35
```
I'm getting this error:
```
paul@extra:~/kokkos/build$ make -j1 VERBOSE=1
/usr/local/bin/cmake -S/home/paul/kokkos/kokkos -B/home/paul/kokkos/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/local/bin/cmake -E cmake_progress_start /home/paul/kokkos/build/CMakeFiles /home/paul/kokkos/build//CMakeFiles/progress.marks
make  -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/home/paul/kokkos/build'
make  -f core/src/CMakeFiles/kokkoscore.dir/build.make core/src/CMakeFiles/kokkoscore.dir/depend
make[2]: Entering directory '/home/paul/kokkos/build'
cd /home/paul/kokkos/build && /usr/local/bin/cmake -E cmake_depends "Unix Makefiles" /home/paul/kokkos/kokkos /home/paul/kokkos/kokkos/core/src /home/paul/kokkos/build /home/paul/kokkos/build/core/src /home/paul/kokkos/build/core/src/CMakeFiles/kokkoscore.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/paul/kokkos/build'
make  -f core/src/CMakeFiles/kokkoscore.dir/build.make core/src/CMakeFiles/kokkoscore.dir/build
make[2]: Entering directory '/home/paul/kokkos/build'
[  0%] Building CXX object core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_Stacktrace.cpp.o
cd /home/paul/kokkos/build/core/src && /home/paul/kokkos/kokkos/bin/nvcc_wrapper -Dkokkoscore_EXPORTS -I/home/paul/kokkos/build/core/src -I/home/paul/kokkos/kokkos/core/src -I/home/paul/kokkos/build -fPIC --relocatable-device-code=true -expt-extended-lambda -Wext-lambda-captures-this -expt-relaxed-constexpr -arch=sm_35 -Xcompiler -fopenmp -std=c++17 -o CMakeFiles/kokkoscore.dir/impl/Kokkos_Stacktrace.cpp.o -c /home/paul/kokkos/kokkos/core/src/impl/Kokkos_Stacktrace.cpp
nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
Error: Internal Compiler Error (codegen): "there was an error in verifying the lgenfe output!"
```

**Comment by calewis** [comment]
@pauljurczak this appears to work for me with Cuda 11.1.0 and Cuda 11.0.2 on our end:
My configure script:
```
#! /bin/bash

module load gcc/8.4.0 cuda/11.0.2 cmake/3.13.2

cmake ../../kokkos/ \
  -DCMAKE_CXX_COMPILER="$HOME/Software/kokkos-meta/kokkos/bin/nvcc_wrapper" \
  -DKokkos_ENABLE_OPENMP=ON \
  -DKokkos_ENABLE_CUDA_LAMBDA=ON \
  -DKokkos_ARCH_KEPLER35=ON \
  -DKokkos_ENABLE_CUDA=ON \
  -DCMAKE_INSTALL_PREFIX="../../install/kepler-gcc" \
  -DKokkos_CXX_STANDARD=17 \
  -DCMAKE_CXX_STANDARD=17
```
Results of configure:
```
-- The CXX compiler identification is GNU 8.4.0
-- Check for working CXX compiler:$HOME/Software/kokkos-meta/kokkos/bin/nvcc_wrapper
-- Check for working CXX compiler: $HOME/Software/kokkos-meta/kokkos/bin/nvcc_wrapper -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Setting build type to 'RelWithDebInfo' as none was specified.
-- Setting policy CMP0074 to use <Package>_ROOT variables
-- The project name is: Kokkos
-- Compiler Version: 11.0.194
-- Using -std=c++17 for C++17 standard as feature
-- Built-in Execution Spaces:
--     Device Parallel: Kokkos::Cuda
--     Host Parallel: Kokkos::OpenMP
--       Host Serial: NONE
--
-- Architectures:
--  KEPLER35
-- Found CUDAToolkit: $SPACK_PATH/spack/linux-centos7-x86_64/gcc-4.8.5/cuda-11.0.2-daywqdi3tzxs7nddo4ppikh6pngvi75g/include (found version "11.0.194")
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE
-- Found TPLCUDA: TRUE
-- Found TPLLIBDL: /usr/lib64/libdl.so
-- Kokkos Devices: OPENMP;CUDA, Kokkos Backends: OPENMP;CUDA
-- Configuring done
-- Generating done
-- Build files have been written to: $HOME/Software/kokkos-meta/build/kepler-gcc8.4
```
Then `make VERBOSE=1` compiles Kokkos_Stacktrace.cpp:
```
[ 60%] Building CXX object core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_Stacktrace.cpp.o
cd Software/kokkos-meta/build/kepler-gcc8.4/core/src && Software/kokkos-meta/kokkos/bin/nvcc_wrapper  -DKOKKOS_DEPENDENCE -I/Software/kokkos-meta/build/kepler-gcc8.4/core/src -I/Software/kokkos-meta/kokkos/core/src -I/Software/kokkos-meta/build/kepler-gcc8.4 -isystem /opt/spack/c7.7/compilers-20200612/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/cuda-11.0.2-daywqdi3tzxs7nddo4ppikh6pngvi75g/include  -O2 -g -DNDEBUG   -expt-extended-lambda -Wext-lambda-captures-this -arch=sm_35 -Xcompiler -fopenmp -std=c++17 -o CMakeFiles/kokkoscore.dir/impl/Kokkos_Stacktrace.cpp.o -c /Software/kokkos-meta/kokkos/core/src/impl/Kokkos_Stacktrace.cpp
nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
```
This compile also works with Cuda 11.1.0.  So at the moment I am not sure how to reproduce this issue.

**Comment by calewis** [comment]
@pauljurczak which commit of Kokkos are you trying to compile.  I did the above with cf0024250e8b9729bb07673e722b3661d308b972

**Comment by pauljurczak** [comment]
Originally, I used the `master` branch. I tried again with `develop` branch and build went fine. Should I avoid master branch in the future?

**Comment by calewis** [comment]
I don't think we are suggesting that people don't use master.  I'll test with Master and see what happens.

So I built the exact same way above with 03ded4e816da5927b946065b005ebfdf38e5a438 and I did not get an error.

**Comment by pauljurczak** [comment]
I added a few CMake options and I'm getting the same error on `develop` branch now. Here is a screenshot of `cmake-gui` with relevant options:

![kokkos1](https://user-images.githubusercontent.com/4933032/96786654-d5131680-13ad-11eb-8a07-ce17d1a16e2e.png)

**Comment by pauljurczak** [comment]
This is a better format of my settings:
```
Commandline options:
-DKokkos_ENABLE_TESTS:BOOL="0" -DBUILD_SHARED_LIBS:BOOL="1" -DKokkos_ENABLE_OPENMP:BOOL="1" -DKokkos_ENABLE_CUDA_CONSTEXPR:BOOL="1" -DKokkos_ENABLE_CUDA:BOOL="1" -DKokkos_ENABLE_CUDA_LAMBDA:BOOL="1" -DKokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE:BOOL="0" -DKokkos_CXX_STANDARD:STRING="17" -DCMAKE_BUILD_TYPE:STRING="Release" -DKokkos_ARCH_KEPLER35:BOOL="1" -DKokkos_ENABLE_CUDA_UVM:BOOL="1" -DKokkos_ENABLE_CUDA_LDG_INTRINSIC:BOOL="1" -DKokkos_ENABLE_EXAMPLES:BOOL="1" -DBUILD_TESTING:BOOL="1" 
```

**Comment by crtrott** [comment]
I was able to reproduce the issue. If you turn off the CUDA_CONSTEXPR (or rather not turn it on explicitly since the default is off) it passes. Note for nvcc this option is the -expt-relaxed-constexpr, where the "-expt" stands for experimental. This also works with the relaxed constexpr stuff on if you drop the standard to C++14. So its a combination of relaxed constexp with C++17. Unfortunately we can't test every possible combination of options so this isn't part of our testing suite right now. We can check whether we can work around the issue (again note to other developers: the defining things are: NVCC 11 + C++17 + Kokkos_ENABLE_CUDA_CONSTEXPR). We also reported the issue to NVIDIA,

**Comment by pauljurczak** [comment]
Not a problem. I don't have a specific need for CUDA_CONSTEXPR. I just turned on all CUDA options the configuration allowed.

**Comment by crtrott** [comment]
Oh you shouldn't turn on Kokkos_ENABLE_CUDA_UVM. Its a bit of a misnomer, but basically it makes the default memory space i.e. what you get when you create Kokkos::View<double*> to UVM memory. Which can be useful (i.e. you don't need to worry about data movement) but it can cause performance issues.

**Comment by Char-Aznable** [comment]
I have hit this bug as well with cuda 11.1.74 and gcc 9.3.0 when compiling this:
```c++
cd /home/aznb/mycodes/testKokkos/build_cuda/_deps/kokkos-build/core/src && /home/aznb/spack/var/spack/environments/trilinos-build/.spack-env/view/bin/nvcc_wrapper -DKOKKOS_DEPENDENCE -Dkokkoscore_EXPORTS -I/home/aznb/mycodes/testKokkos/build_cuda/_deps/kokkos-build/core/src -I/home/aznb/mycodes/testKokkos/build_cuda/_deps/kokkos-src/core/src -I/home/aznb/mycodes/testKokkos/build_cuda/_deps/kokkos-build -isystem /home/aznb/spack/var/spack/environments/trilinos-build/.spack-env/view/include -O3 -fopenmp --expt-relaxed-constexpr --expt-extended-lambda -fPIC -march=core-avx2 -mtune=core-avx2 -expt-extended-lambda -Wext-lambda-captures-this -arch=sm_52 -Xcompiler -fopenmp -std=c++17 -o CMakeFiles/kokkoscore.dir/impl/Kokkos_Stacktrace.cpp.o -c /home/aznb/mycodes/testKokkos/build_cuda/_deps/kokkos-src/core/src/impl/Kokkos_Stacktrace.cpp
```

**Comment by crtrott** [comment]
As described above: this issue is a known bug in CUDA11 which happens with -expt-relaxed-constexpr and C++17 on at the same time. It seems to be related to exception handling but isn't fully clear. It is unlikely that we can fix this reliably on our end.

So: with CUDA 11 do not turn on C++17 and Kokkos_ENABLE_CUDA_CONSTEXPR at the same time.

**Comment by pgrete** [comment]
FYI for people still having/watching this issue: This seems to be fixed with Cuda 11.2 (tried on Power9 and x86 with gcc host compilers).
